src/execution/index/art/node4.cpp

```diff
@@ -86,45 +97,69 @@ void Node4::InsertChild(Node *&node, uint8_t key_byte, Node *new_child) {
 		n->key[pos] = key_byte;
 		n->children[pos] = new_child;
 		n->count++;
+
 	} else {
-		// Grow to Node16
+		// node is full, grow to Node16
 		auto new_node = Node16::New();
-		new_node->count = 4;
+		art.memory_size += new_node->MemorySize(art, false);
+		new_node->count = n->count;
 		new_node->prefix = std::move(node->prefix);
-		for (idx_t i = 0; i < 4; i++) {
+
+		for (idx_t i = 0; i < n->count; i++) {
 			new_node->key[i] = n->key[i];
 			new_node->children[i] = n->children[i];
 			n->children[i] = nullptr;
 		}
-		// Delete old node and replace it with new Node16
+		n->count = 0;
+
+		D_ASSERT(art.memory_size >= node->MemorySize(art, false));
+		art.memory_size -= node->MemorySize(art, false);
 		Node::Delete(node);
 		node = new_node;
-		Node16::InsertChild(node, key_byte, new_child);
+		Node16::InsertChild(art, node, key_byte, new_child);
 	}
 }
 
-void Node4::EraseChild(Node *&node, int pos, ART &art) {
+void Node4::EraseChild(ART &art, Node *&node, idx_t pos) {
+
 	Node4 *n = (Node4 *)node;
 	D_ASSERT(pos < n->count);
+	D_ASSERT(n->count > 1);
+
+	// adjust the ART size
+	if (n->ChildIsInMemory(pos)) {
+		auto child = n->GetChild(art, pos);
+		D_ASSERT(art.memory_size >= child->MemorySize(art, true));
+		art.memory_size -= child->MemorySize(art, true);
+	}
+
 	// erase the child and decrease the count
 	n->children[pos].Reset();
 	n->count--;
+	D_ASSERT(n->count >= 1);
+
 	// potentially move any children backwards
 	for (; pos < n->count; pos++) {
 		n->key[pos] = n->key[pos + 1];
 		n->children[pos] = n->children[pos + 1];
 	}
 	// set any remaining nodes as nullptr
-	for (; pos < 4; pos++) {
+	for (; pos < Node4::GetSize(); pos++) {
 		n->children[pos] = nullptr;
 	}
 
-	// This is a one way node
+	// this is a one way node, compress
 	if (n->count == 1) {
+
+		// get only child and concatenate prefixes
 		auto child_ref = n->GetChild(art, 0);
 		// concatenate prefixes
-		child_ref->prefix.Concatenate(n->key[0], node->prefix);
+		child_ref->prefix.Concatenate(art, n->key[0], node->prefix);
+		// free this node
 		n->children[0] = nullptr;
```

Comment:
Don't we get a reference to this child, that we use in `node = child_ref` later
Does this not cause a heap-use-after-free?